use llvm::SmallVector to reduce allocations

toshok opened the pull request:
using this for the std::vector's we use during calls helps some, with the biggest reduction being -10% on raytrace.

```
         pyston (calibration)                      :    1.0s baseline: 1.0 (+0.1%)
         pyston nbody_med.py                       :    2.1s baseline: 2.2 (-2.2%)
         pyston interp2.py                         :    4.1s baseline: 4.0 (+0.7%)
         pyston raytrace.py                        :    6.7s baseline: 7.6 (-10.8%)
         pyston nbody.py                           :    7.8s baseline: 8.0 (-2.3%)
         pyston fannkuch.py                        :    6.3s baseline: 6.2 (+1.8%)
         pyston chaos.py                           :   20.1s baseline: 20.7 (-3.2%)
         pyston spectral_norm.py                   :   16.8s baseline: 16.5 (+2.1%)
         pyston fasta.py                           :    5.8s baseline: 5.9 (-2.2%)
         pyston pidigits.py                        :    4.2s baseline: 4.3 (-3.0%)
         pyston richards.py                        :    1.8s baseline: 1.8 (-1.5%)
         pyston deltablue.py                       :    2.5s baseline: 2.4 (+4.5%)
         pyston sre_parse_parse.py                 :    1.9s baseline: 1.9 (-0.1%)
         pyston (geomean-f6d7)                     :    5.4s baseline: 5.4 (-1.5%)
```

I need to figure out the best way to get find all the remaining ones.

undingen commented:
You really should add a commit hook which runs ```make format``` :grimacing:

toshok commented:
sigh yeah :)

On Tue, Feb 24, 2015 at 5:00 AM, Marius Wachtler <notifications@github.com>
wrote:

> You really should add a commit hook which runs make format [image:
> :grimacing:]
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/dropbox/pyston/pull/321#issuecomment-75751935>.
>

toshok commented:
and now I have one :smile:

toshok commented:
After cad46f7 (and re-basing the comparison against current master's performance):

```
         pyston (calibration)                      :    1.0s baseline: 1.0 (+0.0%)
         pyston nbody_med.py                       :    2.2s baseline: 2.3 (-6.1%)
         pyston interp2.py                         :    4.1s baseline: 4.2 (-2.2%)
         pyston raytrace.py                        :    7.0s baseline: 7.4 (-5.5%)
         pyston nbody.py                           :    8.0s baseline: 8.5 (-6.1%)
         pyston fannkuch.py                        :    6.1s baseline: 6.3 (-3.7%)
         pyston chaos.py                           :   20.0s baseline: 21.3 (-6.2%)
         pyston spectral_norm.py                   :   14.9s baseline: 15.2 (-2.1%)
         pyston fasta.py                           :    5.6s baseline: 5.7 (-2.0%)
         pyston pidigits.py                        :    4.2s baseline: 4.3 (-0.3%)
         pyston richards.py                        :    1.6s baseline: 1.7 (-0.5%)
         pyston deltablue.py                       :    2.5s baseline: 2.6 (-3.6%)
         pyston sre_parse_parse.py                 :    1.9s baseline: 1.9 (+1.0%)
         pyston (geomean-f6d7)                     :    5.3s baseline: 5.5 (-3.5%)
```

undingen commented on src/core/containers.h:
```diff
@@ -0,0 +1,153 @@
+    }
+    ~SmallStackVector() {
+        if (datap != &data[0])
+            alloc.deallocate(datap, 1);
```
I think the this should be ```alloc.deallocate(data, capacity)```
Also I think we don't destroy the objects inside...

toshok replied:
yup.  Already have that fixed locally (turns out writing code when I should
be sleeping leads to bad code :)

kmod brought up the fact that we should be able to just use
llvm::SmallVector instead of rolling our own, so I'm going to take a stab
at that this week and update this PR.

On Wed, Feb 25, 2015 at 12:01 PM, Marius Wachtler <notifications@github.com>
wrote:

> In src/core/containers.h
> <https://github.com/dropbox/pyston/pull/321#discussion_r25373518>:
>
> > +    }
> > +    SmallStackVector(std::initializer_list<T> init) {
> > +        capacity = 16;
> > +        _size = init.size();
> > +        datap = &data[0];
> > +
> > +        maybe_resize();
> > +
> > +        int i = 0;
> > +        for (auto e : init) {
> > +            alloc.construct(&datap[i++], std::move(e));
> > +        }
> > +    }
> > +    ~SmallStackVector() {
> > +        if (datap != &data[0])
> > +            alloc.deallocate(datap, 1);
>
> I think the this should be alloc.deallocate(data, capacity)
> Also I think we don't destroy the objects inside...
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/dropbox/pyston/pull/321/files#r25373518>.
>

toshok commented:
I looked over the uses of StlCompatAllocator in objmodel.cpp's std::vector's, and they appear unnecessary.  I went ahead and removed SmallStackVector and replaced all uses with llvm::SmallVector.  current benchmarks against `master` are:

```
Saving results as 'fewer-mallocs'
              pyston (calibration)                      :    1.0s baseline: 1.0 (+0.0%)
              pyston interp2.py                         :    4.0s baseline: 3.9 (+1.4%)
              pyston raytrace.py                        :    6.6s baseline: 6.9 (-4.9%)
              pyston nbody.py                           :    7.6s baseline: 7.4 (+2.7%)
              pyston fannkuch.py                        :    6.3s baseline: 6.2 (+2.0%)
              pyston chaos.py                           :   17.3s baseline: 18.6 (-6.7%)
              pyston spectral_norm.py                   :   16.4s baseline: 16.8 (-2.4%)
              pyston fasta.py                           :    5.9s baseline: 5.9 (-0.4%)
              pyston pidigits.py                        :    4.1s baseline: 4.2 (-1.7%)
              pyston richards.py                        :    1.6s baseline: 1.6 (-2.9%)
              pyston deltablue.py                       :    2.4s baseline: 2.5 (-4.2%)
              pyston (geomean-0b9f)                     :    5.6s baseline: 5.7 (-1.7%)
```

with some variability.  A previous run had everything at negative %.

kmod commented on src/runtime/objmodel.cpp:
```diff
@@ -2484,7 +2488,7 @@ Box* callFunc(BoxedFunctionBase* func, CallRewriteArgs* rewrite_args, ArgPassSpe
     }
 
-    std::vector<Box*> varargs;
+    llvm::SmallVector<Box*, 8> varargs;
```
Hmm I think this might need to use an StlCompatAllocator -- what if given_varargs is not a container and generates objects on the fly?  I think then they could only be referenced by varargs?

kmod replied:
We could also potentially disable GC for the duration of `callFunc` :P  There are barely any allocations in here, though I guess we'd have to be careful about the call to pyElements.

kmod commented on src/runtime/objmodel.cpp:
```diff
@@ -2520,12 +2524,12 @@ Box* callFunc(BoxedFunctionBase* func, CallRewriteArgs* rewrite_args, ArgPassSpe
     }
 
-    std::vector<Box*, StlCompatAllocator<Box*>> unused_positional;
+    llvm::SmallVector<Box*, 8> unused_positional;
```
Hmm I feel iffy about this one... I guess it's true that arg1/arg2/arg3/args are all referenced lower than here which I think will force the values to stay tracked, but I'm not sure if we should rely on that.

toshok commented:
I switched the `Box*` vectors in `objmodel.cpp` back to `StlCompatAllocator<>` managed `std::vectors`.  the speedup is less than before for chaos, which makes me think we're doing a lot of calls which are not being rewritten:

```
              pyston (calibration)                      :    1.0s baseline: 1.0 (-0.0%)
              pyston interp2.py                         :    3.8s baseline: 3.9 (-3.1%)
              pyston raytrace.py                        :    6.5s baseline: 6.9 (-5.2%)
              pyston nbody.py                           :    7.5s baseline: 7.4 (+1.3%)
              pyston fannkuch.py                        :    6.2s baseline: 6.2 (+0.5%)
              pyston chaos.py                           :   17.9s baseline: 18.6 (-3.9%)
              pyston spectral_norm.py                   :   16.3s baseline: 16.8 (-3.1%)
              pyston fasta.py                           :    5.9s baseline: 5.9 (-1.6%)
              pyston pidigits.py                        :    4.1s baseline: 4.2 (-2.2%)
              pyston richards.py                        :    1.6s baseline: 1.6 (-3.5%)
              pyston deltablue.py                       :    2.4s baseline: 2.5 (-4.0%)
              pyston (geomean-0b9f)                     :    5.6s baseline: 5.7 (-2.5%)
```

There's plenty of std::vector usage left (in the rewriter, assembler, and elsewhere) that could be converted to llvm::SmallVector, but we can open an issue (labeled 'Easy', to get some familiar with working with the codebase, running benchmarks, etc) for switching the rest over.

kmod commented:
Cool :)  feel free to open the issue.

As an alternative to patching llvm::SmallVector to support a custom allocator, I wonder if we could somehow register a GC handler to scan a non-gc-managed llvm::SmallVector?  Might be a bit more convoluted but could maybe be easier and more general than patching llvm.
